dotnet/src/SemanticKernel.Abstractions/Filters/Function/IFunctionFilter.cs

```diff
@@ -1,24 +1,22 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
 using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
 
 namespace Microsoft.SemanticKernel;
 
+#pragma warning disable CA1716 // Identifiers should not match keywords (FunctionInvocationCallback next)
+
 /// <summary>
 /// Interface for filtering actions during function invocation.
 /// </summary>
 [Experimental("SKEXP0001")]
 public interface IFunctionFilter
 {
     /// <summary>
-    /// Method which is executed before function invocation.
-    /// </summary>
-    /// <param name="context">Data related to function before invocation.</param>
-    void OnFunctionInvoking(FunctionInvokingContext context);
-
-    /// <summary>
-    /// Method which is executed after function invocation.
+    /// Method which is called asynchronously before function invocation.
     /// </summary>
-    /// <param name="context">Data related to function after invocation.</param>
-    void OnFunctionInvoked(FunctionInvokedContext context);
+    /// <param name="context">Instance of <see cref="FunctionInvocationContext"/> with function invocation details.</param>
+    /// <param name="next">Delegate to the next filter in pipeline or function itself. If it's not invoked, next filter or function won't be invoked.</param>
+    Task OnFunctionInvocationAsync(FunctionInvocationContext context, FunctionInvocationCallback next);
```

Comment:
Why is FunctionInvocationCallback needed, rather than just having that be an IFunctionFilter?

Reply:
> If it's to be able to handle the final step where the actual function is invoked

That is an answer, we also need to be able to wrap function invocation in `try/catch` block to handle exceptions.

> that can be done by just implementing the interface with a private type

Not sure I understand this approach, could you please provide more details?

Reply:
> Not sure I understand this approach, could you please provide more details?

```C#
private sealed class TerminalFilter(Kernel kernel, KernelFunction function) : IFunctionFilter
{
    public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, FunctionInvocationCallback next)
    {
        Debug.Assert(next is null);
        context.Result = await function.InvokeAsync(kernel, context.Arguments, cancellationToken);
    }
}
...
IFunctionFilter last = new TerminalFilter(kernel, this);

Reply:
@stephentoub So, in case of having `IFunctionFilter` as `next` parameter instead of `FunctionInvocationCallback`, the usage will look like this:
```csharp
public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, IFunctionFilter next)
{
    // Perform some actions before function invocation
    await next.OnFunctionInvocationAsync(context, ???);
    // Perform some actions after function invocation
}
```

First of all, users will have to call `OnFunctionInvocationAsync` method manually, which is more complicated rather than do just `next(context)`. The other thing is that users will have to pass something as second parameter (question marks in code example) to this method (which should be reference to the next filter in pipeline or function itself. Maybe it will be possible to create extension method for IFunctionFilter interface that accepts only context, but I will need to double check if that will work. 

The other thing is that with private function filter that performs function invocation (e.g. `TerminalFilter` from your sample) - it will be visible in stack trace of exception, which may be confusing, especially when filters weren't registered at all. It shouldn't be a big problem though.

I guess my main concern is that `IFunctionFilter` as second parameter looks more complicated than callback. What do you think?

Reply:
That's fair. Do we need the dedicated delegate type, or can it just use a Func?

Reply:
> Do we need the dedicated delegate type, or can it just use a Func?

Just `Func` should work. I decided to choose dedicated delegate type to follow the same pattern as in ASP.NET filters just to be consistent, but `Func` should be fine. I will experiment with it.

Reply:
Name of dedicated delegate type looks a little bit shorter than `Func<FunctionInvocationContext, Task>`, but I guess it's the only advantage over `Func` in this case. So, I updated `IFunctionFilter` interface to accept `Func` instead of dedicated delegate type.